library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs

```diff
@@ -423,71 +447,124 @@ onPairGen1 op f =
     instr2 : instr1 : _ -> do
-      -- May be the wrong order?
-      let instrs = [instr2, instr1]
-      res <-
-        if  | allConstants (val <$> instrs) ->
-              let Env.Constant i1 = val instr1
-                  Env.Constant i2 = val instr2
-               in pure (f i1 i2)
-            | otherwise -> do
-              traverse_ addExpanded instrs
-              -- add when we normalize
-              -- copyAndDrop 2
-              addInstr op
-              pure Env.Nop
-      -- remove when we normalize
+      -- Get the result of the operation on the arguments
+      res <- applyOperation instr2 instr1
+      -- drop the two arguments on the stack
       modify @"stack" (VStack.drop 2)
+      -- add the result to the stack
       consVal res typ
+      -- return the operation
       pure res
     _ -> throw @"compilationError" Types.NotEnoughArguments
 
-onOneArgs :: OnTerm1 m (V.Value' Types.Op) Env.Expanded
-onOneArgs op f typ instrs = do
-  v <- traverse (protect . (inst >=> promoteTopStack)) instrs
+onOneArgGen :: OnTerm1Gen m Protect Env.Expanded
+onOneArgGen applyOperation typ arguments = do
+  -- length of arguments should be 1
+  -- so a reverse to mimic two should not do anything!
+  v <- traverse (protect . (inst >=> promoteTopStack)) arguments
+  -- see comments onTwoArgsGen
   case v of
     instr1 : _ -> do
-      res <-
-        if  | allConstants [val instr1] ->
-              let Env.Constant i1 = val instr1
-               in pure (f i1)
-            | otherwise -> do
-              addExpanded instr1
-              -- copyAndDrop 1
-              addInstr op
-              pure Env.Nop
-      -- remove when we normalize
+      res <- applyOperation instr1
       modify @"stack" (VStack.drop 1)
       consVal res typ
       pure res
     _ -> throw @"compilationError" Types.NotEnoughArguments
 
--- todo remove repeat pattern
--- Cons val here?
-pushConstant :: Env.Reduction m => Types.Type -> [Types.RawTerm] -> m Env.Expanded
-pushConstant typ instrs = do
-  v <- traverse (inst >=> promoteTopStack) instrs
-  res <- case v of
-    Env.Constant _ : _ ->
-      pure Env.Nop
-    instr1 : _ ->
-      pure instr1
-    _ -> throw @"compilationError" Types.NotEnoughArguments
-  modify @"stack" (VStack.drop 1)
-  consVal res typ
-  pure res
+------------------------------------------------------------
+-- Functions to call from appliers/reducers above
+------------------------------------------------------------
+
+onTwoArgsNoConst ::
+  Env.Reduction m => Instr.ExpandedOp -> Types.Type -> [Types.RawTerm] -> m Env.Expanded
+onTwoArgsNoConst op =
+  onTwoArgsGen (noTwoConstantCase op)
+
+onOneArgsNoConst ::
+  Env.Reduction m => Instr.ExpandedOp -> Types.Type -> [Types.RawTerm] -> m Env.Expanded
+onOneArgsNoConst op =
+  onOneArgGen (noOneConstantCase op)
+
+onTwoArgs :: OnTerm2 m (V.Value' Types.Op) Env.Expanded
+onTwoArgs op f =
+  onTwoArgsGen
+    ( \instr2 instr1 -> do
+        -- May be the wrong order?
```

Comment:
why does it matter? we're just calling `allConstants`, no?

Reply:
I don't really see why we need to make a temporary list here at all tbh

Reply:
For abstraction purposes we did this in the past, as we send it into the all check function... basically to scale it up to N arguments with the same pattern... since haskell's type system makes it hard to abstract